core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -817,16 +816,40 @@ class PartKeyLuceneIndex(ref: DatasetRef,
     logger.info(s"Refreshed index searchers to make reads consistent for dataset=$ref shard=$shardNum")
   }
 
-  //scalastyle:off method.length
+  val regexChars = Array('.', '?', '+', '*', '|', '{', '}', '[', ']', '(', ')', '"', '\\')
+  val regexCharsMinusPipe = (regexChars.toSet - '|').toArray
+
+  // scalastyle:off method.length
   private def leafFilter(column: String, filter: Filter): Query = {
+
+    def equalsQuery(value: String): Query = {
+      if (value.nonEmpty) new TermQuery(new Term(column, value))
+      else leafFilter(column, NotEqualsRegex(".+")) // value="" means the label is absent or has an empty value.
+    }
+
     filter match {
       case EqualsRegex(value) =>
         val regex = removeRegexAnchors(value.toString)
-        if (regex.replaceAll("\\.\\*", "") == "")
+        if (regex == "") {
+          // if label=~"" then match empty string or label not present condition too
+          leafFilter(column, NotEqualsRegex(".+"))
+        } else if (regex.replaceAll("\\.\\*", "") == "") {
+          // if label=~".*" then match all docs since promQL matches .* with absent label too
           new MatchAllDocsQuery
-        else if (regex.nonEmpty)
+        } else if (regex.forall(c => !regexChars.contains(c))) {
+          // if all regex special chars absent, then treat like Equals
+          equalsQuery(regex)
+        } else if (regex.forall(c => !regexCharsMinusPipe.contains(c))) {
+          // if pipe is only regex special char present, then convert to IN query
+          new TermInSetQuery(column, regex.split('|').map(t => new BytesRef(t)): _*)
+        } else if (regex.endsWith(".*") && regex.length > 2 &&
```

Comment:
should we also handle cases like ".*.*"

Reply:
`..` would not be a prefix query. So it should not match for this condition

Reply:
sorry the \* was not escaped.
I mean we can drop all the .\* on the right side.

Reply:
Although possible, it unlikely to get any or too many of these cases. I am leaning towards keeping code simple to maintain. The fallback always works functionally